Initial build-using-dockerfile implementation

nalind opened the pull request:
Add a `build-using-dockerfile` command which uses `openshift/imagebuilder` to wrap parsing and handle most of the dispatching, and the logic behind the other subcommands to implement assorted instructions.

rhatdan commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+			Usage: "`tag` to apply to the built image",
+		},
+		cli.StringSliceFlag{
+			Name:  "f",
```
Name:  "file, f",

rhatdan replied:
Do we want this or do we just specify Dockerfile in the command.

buildah bud Dockerfile  
Or do you want to match docker syntax totally>

nalind replied:
Not totally, but I think we at least need this flag to handle cases where a directory that we're pointed to as a build context is one of those where we're expected to able to select among multiple Dockerfiles with names like Dockerfile.${something}.

rhatdan commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+			Usage: "`pathname` of a Dockerfile",
+		},
+	}
+	buildDescription = "Builds an OCI image using instructions in one or more Dockerfiles."
```
Does this accept more then one Dockerfiles?  Or is this a copy/paste error.

nalind replied:
Traditional imagebuilder (which I'd like to reuse some of this logic in when we get there) does, to a degree: the FROM instructions in files after the first are effectively ignored, but other instructions are not, and the image is committed after instructions from all files have been executed.  I'll clean up the logic to mimic this.

rhatdan commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+		Description: buildDescription,
+		Flags:       buildFlags,
+		Action:      buildCmd,
+		ArgsUsage:   "CONTEXT-DIRECTORY",
```
In Docker build this is a PATH-or-URL

nalind replied:
Support for remote context archives or Dockerfiles wasn't in there.  Adding.

rhatdan commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+func buildCmd(c *cli.Context) error {
+	output := ""
+	if c.IsSet("t") {
+		output = c.String("t")
```
tag

nalind replied:
👍

rhatdan commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+		quiet = c.Bool("quiet")
+	}
+	dockerfiles := []string{}
+	if c.IsSet("f") {
```
file

nalind replied:
👍

nalind commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(8) [Buildah] Add "dockah build-using-dockerfile" aka "buildah bud"](https://trello.com/c/AyBXqoRY/443-8-buildah-add-dockah-build-using-dockerfile-aka-buildah-bud)

rhatdan commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+		Output:              output,
+	}
+
+	return imagebuildah.BuildDockerfileFromFile(store, options, dockerfiles...)
```
Should this just be BuildDockerfile(...)

TomSweeneyRedHat replied:
👍

TomSweeneyRedHat replied:
👍

rhatdan commented on imagebuildah/build.go:
```diff
@@ -0,0 +1,331 @@
+	systemContext                  *types.SystemContext
+}
+
+func getSystemContext(signaturePolicyPath string) *types.SystemContext {
```
Should this function be just
systemContext(...)

TomSweeneyRedHat replied:
fwiw, I prefer keeping the get in the name here.

rhatdan replied:
I am fine with get, but I think @mrunalp pointed out that there is some kind of GOLANG standard that says this is discouraged.  I think golint and goverify used in CRI-O and runc give you errors when you start functions with get.

mrunalp replied:
Here is the suggested style reference: https://golang.org/doc/effective_go.html#Getters

nalind replied:
Hmm, that's focused more on wrappers for reading and writing object members, but in terms of what it does, it probably makes more sense to call it makeSystemContext().

rhatdan commented on imagebuildah/build.go:
```diff
@@ -0,0 +1,331 @@
+	return sc
+}
+
+func (b *buildahExecutor) Preserve(path string) error {
```
We need to add comments to all externally defined functions.  We need to start running golint and govalidate against our code.  I have no idea what Preserve does and it seems to be not implemented yet.

nalind replied:
See #62 for adding formatting and style checking to CI.  The `Preserve` method is used to signal the `Executor` that the path has been marked as part of a volume, which means we need to keep instructions other than an explicit ADD or COPY from modifying anything below that point.  The imagebuilder dockerclient implementation uses caching/restoring around its handling of RUN instructions to accomplish this.  Adding that.

rhatdan commented on imagebuildah/build.go:
```diff
@@ -0,0 +1,331 @@
+	return nil
+}
+
+func (b *buildahExecutor) Copy(excludes []string, copies ...imagebuilder.Copy) error {
```
Comment.

rhatdan commented on imagebuildah/build.go:
```diff
@@ -0,0 +1,331 @@
+	return nil
+}
+
+func (b *buildahExecutor) Run(run imagebuilder.Run, config docker.Config) error {
```
Comment

rhatdan commented:
BTW This is pretty awesome.

nalind commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+package main
```
bud's a lot less typing, so I'll rename it to that.

nalind commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+var (
+	buildFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "quiet",
```
👍

nalind commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+			Name:  "pull",
+			Usage: "pull the image if not present",
+		},
+		cli.BoolFlag{
```
This defaults to true, and --pull-always defaults to false, like "buildah from" does.  I'll note that "docker build"'s --pull is equivalent to --pull-always, so maybe we need to drop --pull from both, and rename --pull-always to --pull?

TomSweeneyRedHat commented on imagebuildah/build.go:
```diff
@@ -0,0 +1,334 @@
+}
+
+// BuildDockerfile builds an image using instructions from the named file.
+func BuildDockerfileFromFile(store storage.Store, options BuildDockerfileOptions, dockerfile ...string) error {
```
If you change the name to BuildDockerfileFromFile in build.go, you probably want to do the same here.

nalind commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+		},
+		cli.StringSliceFlag{
+			Name:  "arg",
+			Usage: "`argument=value` to supply to the builder",
```
This is used to supply a value for a variable named by an ARG instruction in a Dockerfile, after which the variable's value can be referred to in other instructions, and will be added to the environment list we get when we're told to process a RUN instruction.  Renaming this to `--build-arg`.

nalind commented on cmd/buildah/build.go:
```diff
@@ -0,0 +1,136 @@
+			Usage: "`argument=value` to supply to the builder",
+		},
+		cli.StringFlag{
+			Name:  "t",
```
👍

nalind commented:
Updated:
* rebased
* golint clean
* tests no longer fail without #61 
* `Preserve()` is now implemented, so handling of volumes during bud should be more in line with expectations
* the bud command's `--arg` flag is now named `--build-arg`
* the bud command's `-t`, `-f`, and `-q` flags now also have long names `--tag`, `--file`, and `--quiet`
* Dockerfiles can be specified using http/https URLs
* context directories can be specified as "git://" URLs, "github.com/..." repositories, or as http/https URLs of archives
* handling of multiple Dockerfiles is more in line with how `openshift/imagebuildah` uses them (in order, only the first FROM: counts)
* renamed `getSystemContext()` to `makeSystemContext()`
* more tests

rhatdan commented:
Lets remove the 1.6 tests, in a separate patch.

rhatdan commented:
LGTM

Do we eventually want to turn ImageBuilder to use Buildah?  Might want to move this code out of ImageBuilder into Buildah at that point.

nalind commented:
Moving the tell-Travis-to-stop-using-1.6 patch to #65 and rebasing before merging.

nalind commented:
I think the imagebuilder CLI could be taught to toggle between its current dockerclient Executor and the one we implement here, as @smarterclayton suggested in https://github.com/openshift/imagebuilder/issues/33.  We don't yet support all of the options and hooks that the imagebuilder CLI tool wants to be able to specify, but that's fixable.

smarterclayton commented:
You should also be able to snip more of the dependency tree that we pull in - let me know if there are changes that need to be done to make the core imagebuilder code not depend on docker dependencies.

rhatdan commented on cmd/buildah/bud.go:
```diff
@@ -0,0 +1,176 @@
+	budFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "quiet, q",
+			Usage: "refrain from announcing build instructions",
```
Should we reverse this and change it to verbose, and then default it to value so we don't announce build instructions?

rhatdan commented:
man pages and bash completions?

nalind commented:
Added a man page and updated the bash-completion script.

rhatdan commented on docs/buildah-bud.md:
```diff
@@ -0,0 +1,79 @@
+buildah bud - Build an image using instructions from Dockerfiles.
+
+## SYNOPSIS
+**buildah** **bud | build-using-dockerfile** [*options* [...]] [**context**]
```
Is the context directory optional?

rhatdan replied:
I guess it is if you specify a Dockerfile.

nalind replied:
If at least one of the Dockerfiles is a local file, we assume the directory that it's in is the context directory.

rhatdan commented:
@rh-atomic-bot r+ 390defd
